### PR TITLE
Feat/mock signing revamp

### PR DESCRIPTION
**Pull request by jferrant**
Closes https://github.com/stacks-network/stacks-core/issues/5056

### CODE REVIEW

File: stackslib/src/net/stackerdb/mod.rs [hunk 1]
```diff
@@ -152,7 +152,7 @@ pub const STACKERDB_MAX_PAGE_COUNT: u32 = 2;
 
 pub const STACKERDB_SLOTS_FUNCTION: &str = "stackerdb-get-signer-slots";
 pub const STACKERDB_CONFIG_FUNCTION: &str = "stackerdb-get-config";
-pub const MINER_SLOT_COUNT: u32 = 2;
+pub const MINER_SLOT_COUNT: u32 = 3;
```
Comment (jcnelson) [inline stackslib/src/net/stackerdb/mod.rs:155]: I don't think you need to change this?  Can we just use slot 0/1?

File: testnet/stacks-node/src/neon_node.rs [hunk 2]
```diff
@@ -2255,6 +2262,151 @@ impl BlockMinerThread {
...
+        let miner_config = self.config.get_miner_config();
+        if !miner_config.pre_nakamoto_miner_messaging {
+            debug!("Pre-Nakamoto mock miner messaging is disabled");
+            return Ok(());
+        }
+
+        let burn_db_path = self.config.get_burn_db_file_path();
+        let burn_db = SortitionDB::open(&burn_db_path, false, self.burnchain.pox_constants.clone())
+            .expect("FATAL: could not open sortition DB");
+        let p2p_net = StacksNode::setup_peer_network(
```
Comment (jcnelson) [inline testnet/stacks-node/src/neon_node.rs:2276]: I don't think you need to go so far as to instantiate a whole other `PeerNetwork`.

File: testnet/stacks-node/src/neon_node.rs [hunk 3]
```diff
@@ -2255,6 +2262,151 @@ impl BlockMinerThread {
...
+            &self.config,
+            &self.config.atlas,
+            self.burnchain.clone(),
+        );
+        let epoch_id = p2p_net.get_current_epoch().epoch_id;
```
Comment (jcnelson) [inline testnet/stacks-node/src/neon_node.rs:2281]: You can also get this from the sortition DB.

File: testnet/stacks-node/src/neon_node.rs [hunk 4]
```diff
@@ -2255,6 +2262,151 @@ impl BlockMinerThread {
...
+            }
+        }
+
+        let server_version = version_string(
+            "stacks-node",
+            option_env!("STACKS_NODE_VERSION")
+                .or(option_env!("CARGO_PKG_VERSION"))
+                .unwrap_or("0.0.0.0"),
+        );
+        let stacks_tip_height = p2p_net.stacks_tip.height;
```
Comment (jcnelson) [inline testnet/stacks-node/src/neon_node.rs:2353]: This information -- `stacks_tip`, `burnchain_tip`, and `chain_view` -- can be synthesized through queries against the Stacks chainstate DB and sortition DB.

**Review by jcnelson — COMMENTED** [review]
Overall this looks good, but I think you can avoid creating `PeerNetwork` over and over again.  I also don't think it's safe to change the number of miner slots -- nodes today already think that there are two slots, so nodes that don't run this version will not sync their `.miners` stackerdb with nodes that do.

File: testnet/stacks-node/src/neon_node.rs (hunk 2 above)
Comment (jferrant) [reply testnet/stacks-node/src/neon_node.rs]: Yes this was a temp. have replaced!

File: stackslib/src/net/stackerdb/mod.rs (hunk 1 above)
Comment (jferrant) [reply stackslib/src/net/stackerdb/mod.rs:155]: Yep! Reverted.

**Review by jcnelson — DISMISSED** [review]
LGTM!

File: stacks-signer/src/v0/signer.rs [hunk 5]
```diff
@@ -140,6 +139,25 @@ impl SignerTrait<SignerMessage> for Signer {
...
                         }
+                        SignerMessage::MockProposal(mock_proposal) => {
+                            let epoch = match stacks_client.get_node_epoch() {
+                                Ok(epoch) => epoch,
+                                Err(e) => {
+                                    warn!("{self}: Failed to determine node epoch. Cannot mock sign: {e}");
+                                    return;
+                                }
+                            };
+                            debug!("{self}: received a mock block proposal.";
```
Comment (obycode) [inline stacks-signer/src/v0/signer.rs:150]: This might be useful as an `info!`.

File: stacks-signer/src/v0/signer.rs [hunk 6]
```diff
@@ -140,6 +139,25 @@ impl SignerTrait<SignerMessage> for Signer {
...
+                                "current_reward_cycle" => current_reward_cycle,
+                                "epoch" => ?epoch
+                            );
+                            if epoch == StacksEpochId::Epoch25
+                                && self.reward_cycle == current_reward_cycle
+                            {
+                                // We are in epoch 2.5, so we should mock mine to prove we are still alive.
```
Comment (obycode) [inline stacks-signer/src/v0/signer.rs:157]:
```suggestion
                                // We are in epoch 2.5, so we should mock sign to prove we are still alive.
```

File: testnet/stacks-node/src/neon_node.rs [hunk 7]
```diff
@@ -3595,7 +3794,12 @@ impl RelayerThread {
         if let Ok(miner_handle) = thread::Builder::new()
             .name(format!("miner-block-{}", self.local_peer.data_url))
             .stack_size(BLOCK_PROCESSOR_STACK_SIZE)
-            .spawn(move || miner_thread_state.run_tenure())
+            .spawn(move || {
+                if let Err(e) = miner_thread_state.send_mock_miner_messages() {
+                    warn!("Failed to send mock miner messages: {}", e);
+                }
```
Comment (obycode) [inline testnet/stacks-node/src/neon_node.rs:3800]: Should we run the tenure first, then do this mocking?

Comment (jferrant) [reply testnet/stacks-node/src/neon_node.rs:3800]: I don't think it makes a difference either way. I can try moving it to after and see if things break (I think running the tenure does change internal state so not sure things will be instantiated still as they should be.)

Comment (jferrant) [reply testnet/stacks-node/src/neon_node.rs:3800]: K I don't understand why, but putting it after breaks things as it fails to calculate the reward set in this case as expected. ... I am sure it is some other dtate issue that manifests down the line as this, but leaving as is for now to get this in asap.

**Review by obycode — DISMISSED** [review]

**Review by hstove — DISMISSED** [review]
LGTM!

**Review by obycode — DISMISSED** [review]

**Review by hstove — DISMISSED** [review]

**Review by jcnelson — DISMISSED** [review]

**Review by hstove — APPROVED** [review]

**Review by obycode — APPROVED** [review]

**Review by hstove — APPROVED** [review]

**Pull request merged**